File: docs/source/src/rust/user-guide/getting-started.rs
fn main() -> Result<(), Box<dyn std::error::Error>> {
    // --8<-- [start:df]
    use chrono::prelude::*;
    use polars::prelude::*;

    let mut df: DataFrame = df!(
        "name" => ["Alice Archer", "Ben Brown", "Chloe Cooper", "Daniel Donovan"],
        "birthdate" => [
            NaiveDate::from_ymd_opt(1997, 1, 10).unwrap(),
            NaiveDate::from_ymd_opt(1985, 2, 15).unwrap(),
            NaiveDate::from_ymd_opt(1983, 3, 22).unwrap(),
            NaiveDate::from_ymd_opt(1981, 4, 30).unwrap(),
        ],
        "weight" => [57.9, 72.5, 53.6, 83.1],  // (kg)
        "height" => [1.56, 1.77, 1.65, 1.75],  // (m)
    )
    .unwrap();
    println!("{}", df);
    // --8<-- [end:df]

    // --8<-- [start:csv]
    use std::fs::File;

    let mut file = File::create("../../../assets/data/output.csv").expect("could not create file");
    CsvWriter::new(&mut file)
        .include_header(true)
        .with_separator(b',')
        .finish(&mut df)?;
    let df_csv = CsvReadOptions::default()
        .with_infer_schema_length(None)
        .with_has_header(true)
        .with_parse_options(CsvParseOptions::default().with_try_parse_dates(true))
        .try_into_reader_with_file_path(Some("../../../assets/data/output.csv".into()))?
        .finish()?;
    println!("{}", df_csv);
    // --8<-- [end:csv]

    // --8<-- [start:select]
    let result = df
        .clone()
        .lazy()
        .select([
            col("name"),
            col("birthdate").dt().year().alias("birth_year"),
            (col("weight") / col("height").pow(2)).alias("bmi"),
        ])
        .collect()?;
    println!("{}", result);
    // --8<-- [end:select]

    // --8<-- [start:expression-expansion]
    let result = df
        .clone()
        .lazy()
        .select([
            col("name"),
            (cols(["weight", "height"]) * lit(0.95))
                .round(2)
                .name()
                .suffix("-5%"),
        ])
        .collect()?;
    println!("{}", result);
    // --8<-- [end:expression-expansion]

    // --8<-- [start:with_columns]
    let result = df
        .clone()
        .lazy()
        .with_columns([
            col("birthdate").dt().year().alias("birth_year"),
            (col("weight") / col("height").pow(2)).alias("bmi"),
        ])
        .collect()?;
    println!("{}", result);
    // --8<-- [end:with_columns]

    // --8<-- [start:filter]
    let result = df
        .clone()
        .lazy()
        .filter(col("birthdate").dt().year().lt(lit(1990)))
        .collect()?;
    println!("{}", result);
    // --8<-- [end:filter]

    // --8<-- [start:filter-multiple]
    let result = df
        .clone()
        .lazy()
        .filter(
            col("birthdate")
                .is_between(
                    lit(NaiveDate::from_ymd_opt(1982, 12, 31).unwrap()),
                    lit(NaiveDate::from_ymd_opt(1996, 1, 1).unwrap()),
                    ClosedInterval::Both,
                )
                .and(col("height").gt(lit(1.7))),
        )
        .collect()?;
    println!("{}", result);
    // --8<-- [end:filter-multiple]

    // --8<-- [start:group_by]
    // Use `group_by_stable` if you want the Python behaviour of `maintain_order=True`.
    let result = df
        .clone()
        .lazy()
        .group_by([(col("birthdate").dt().year() / lit(10) * lit(10)).alias("decade")])
        .agg([len()])
        .collect()?;
    println!("{}", result);
    // --8<-- [end:group_by]

    // --8<-- [start:group_by-agg]
    let result = df
        .clone()
        .lazy()
        .group_by([(col("birthdate").dt().year() / lit(10) * lit(10)).alias("decade")])
        .agg([
            len().alias("sample_size"),
            col("weight").mean().round(2).alias("avg_weight"),
            col("height").max().alias("tallest"),
        ])
        .collect()?;
    println!("{}", result);
    // --8<-- [end:group_by-agg]

    // --8<-- [start:complex]
    let result = df
        .clone()
        .lazy()
        .with_columns([
            (col("birthdate").dt().year() / lit(10) * lit(10)).alias("decade"),
            col("name").str().split(lit(" ")).list().first(),
        ])
        .select([all().exclude(["birthdate"])])
        .group_by([col("decade")])
        .agg([
            col("name"),
            cols(["weight", "height"])
                .mean()
                .round(2)
                .name()
                .prefix("avg_"),
        ])
        .collect()?;
    println!("{}", result);
    // --8<-- [end:complex]

    // --8<-- [start:join]
    let df2: DataFrame = df!(
        "name" => ["Ben Brown", "Daniel Donovan", "Alice Archer", "Chloe Cooper"],
        "parent" => [true, false, false, false],
        "siblings" => [1, 2, 3, 4],
    )
    .unwrap();

    let result = df
        .clone()
        .lazy()
        .join(
            df2.clone().lazy(),
            [col("name")],
            [col("name")],
            JoinArgs::new(JoinType::Left),
        )
        .collect()?;

    println!("{}", result);
    // --8<-- [end:join]

    // --8<-- [start:concat]
    let df3: DataFrame = df!(
        "name" => ["Ethan Edwards", "Fiona Foster", "Grace Gibson", "Henry Harris"],
        "birthdate" => [
            NaiveDate::from_ymd_opt(1977, 5, 10).unwrap(),
            NaiveDate::from_ymd_opt(1975, 6, 23).unwrap(),
            NaiveDate::from_ymd_opt(1973, 7, 22).unwrap(),
            NaiveDate::from_ymd_opt(1971, 8, 3).unwrap(),
        ],
        "weight" => [67.9, 72.5, 57.6, 93.1],  // (kg)
        "height" => [1.76, 1.6, 1.66, 1.8],  // (m)
    )
    .unwrap();

    let result = concat(
        [df.clone().lazy(), df3.clone().lazy()],
        UnionArgs::default(),
    )?
    .collect()?;
    println!("{}", result);
    // --8<-- [end:concat]

    Ok(())
}

File: docs/source/src/python/user-guide/getting-started.py
# --8<-- [start:df]
import polars as pl
import datetime as dt

df = pl.DataFrame(
    {
        "name": ["Alice Archer", "Ben Brown", "Chloe Cooper", "Daniel Donovan"],
        "birthdate": [
            dt.date(1997, 1, 10),
            dt.date(1985, 2, 15),
            dt.date(1983, 3, 22),
            dt.date(1981, 4, 30),
        ],
        "weight": [57.9, 72.5, 53.6, 83.1],  # (kg)
        "height": [1.56, 1.77, 1.65, 1.75],  # (m)
    }
)

print(df)
# --8<-- [end:df]

# --8<-- [start:csv]
df.write_csv("docs/assets/data/output.csv")
df_csv = pl.read_csv("docs/assets/data/output.csv", try_parse_dates=True)
print(df_csv)
# --8<-- [end:csv]

# --8<-- [start:select]
result = df.select(
    pl.col("name"),
    pl.col("birthdate").dt.year().alias("birth_year"),
    (pl.col("weight") / (pl.col("height") ** 2)).alias("bmi"),
)
print(result)
# --8<-- [end:select]

# --8<-- [start:expression-expansion]
result = df.select(
    pl.col("name"),
    (pl.col("weight", "height") * 0.95).round(2).name.suffix("-5%"),
)
print(result)
# --8<-- [end:expression-expansion]

# --8<-- [start:with_columns]
result = df.with_columns(
    birth_year=pl.col("birthdate").dt.year(),
    bmi=pl.col("weight") / (pl.col("height") ** 2),
)
print(result)
# --8<-- [end:with_columns]

# --8<-- [start:filter]
result = df.filter(pl.col("birthdate").dt.year() < 1990)
print(result)
# --8<-- [end:filter]

# --8<-- [start:filter-multiple]
result = df.filter(
    pl.col("birthdate").is_between(dt.date(1982, 12, 31), dt.date(1996, 1, 1)),
    pl.col("height") > 1.7,
)
print(result)
# --8<-- [end:filter-multiple]

# --8<-- [start:group_by]
result = df.group_by(
    (pl.col("birthdate").dt.year() // 10 * 10).alias("decade"),
    maintain_order=True,
).len()
print(result)
# --8<-- [end:group_by]

# --8<-- [start:group_by-agg]
result = df.group_by(
    (pl.col("birthdate").dt.year() // 10 * 10).alias("decade"),
    maintain_order=True,
).agg(
    pl.len().alias("sample_size"),
    pl.col("weight").mean().round(2).alias("avg_weight"),
    pl.col("height").max().alias("tallest"),
)
print(result)
# --8<-- [end:group_by-agg]

# --8<-- [start:complex]
result = (
    df.with_columns(
        (pl.col("birthdate").dt.year() // 10 * 10).alias("decade"),
        pl.col("name").str.split(by=" ").list.first(),
    )
    .select(
        pl.all().exclude("birthdate"),
    )
    .group_by(
        pl.col("decade"),
        maintain_order=True,
    )
    .agg(
        pl.col("name"),
        pl.col("weight", "height").mean().round(2).name.prefix("avg_"),
    )
)
print(result)
# --8<-- [end:complex]

# --8<-- [start:join]
df2 = pl.DataFrame(
    {
        "name": ["Ben Brown", "Daniel Donovan", "Alice Archer", "Chloe Cooper"],
        "parent": [True, False, False, False],
        "siblings": [1, 2, 3, 4],
    }
)

print(df.join(df2, on="name", how="left"))
# --8<-- [end:join]

# --8<-- [start:concat]
df3 = pl.DataFrame(
    {
        "name": ["Ethan Edwards", "Fiona Foster", "Grace Gibson", "Henry Harris"],
        "birthdate": [
            dt.date(1977, 5, 10),
            dt.date(1975, 6, 23),
            dt.date(1973, 7, 22),
            dt.date(1971, 8, 3),
        ],
        "weight": [67.9, 72.5, 57.6, 93.1],  # (kg)
        "height": [1.76, 1.6, 1.66, 1.8],  # (m)
    }
)

print(pl.concat([df, df3], how="vertical"))
# --8<-- [end:concat]
